## Practice 1

---

For this practice, you will be using Python to do some analysis of the text of Shakespeare's plays.

---


Let's first try running some simple Python code to make sure that everything is set up properly and ready to go.   The code in the next box will open `Shakespeare.txt`, read the file line by line, and tokenize each line that it reads in.    Try running the code by selecting the box and typing `Shift-Return`, i.e., hit the carriage return key while you are holding the shift key.   It may take a few seconds to run.

In [1]:
# this imports the SimpleTokenize function from the simple_tokenize.py file that you uploaded
from simple_tokenize import simple_tokenize

# Now, let's tokenize Shakespeare's plays
filename = 'Shakespeare.txt'
with open(filename) as f:
    for line in f:
        # tokenize, one line at a time
        t = simple_tokenize(line)
t

['the', 'end']

---
### Important

The questions that follow ask you to implement functions whose prototypes are given to you. **Do _not_ change the prototypes of the functions. Do _not_ write code outside of the functions.** In particular, do not use the same cell as the function declaration invoke the function.

You may use specific cells, identified by `# Your tests here`, for test purposes. Code in these cells will *not* be executed when marking your assignment.

---

#### Question 1:
After the code has finished running, the notebook will display the resulting output immediately after cell.   In this case, you should see the output `['the', 'end']`. In the next cell, briefly explain why the code produced this output.


#### Your answer to Question 1:
The simple_tokenize function make our text lowercase and then returns a list of lowercase strings that matches the desired pattern. In our case, the desired pattern is that the string starts with a lowercase letter between a and z(inclusive) and then has an optional apostrophe with some lowercase letters between a and z(inclusive) from behind. 

The result of each line of the text after applying simple_tokenize is stored in a vaiable t which would be rewritten after the next line being processed. The last line in 'Shakespeare.txt' is THE END, so the output should be ['the', 'end'].





Now it is time for you to write some code.   Let's find the most frequently appearing tokens in Shakespeare's work.

#### Question 2:
In the next box, implement the function `top_50_tokens` using Python code that return the list of the 50 most frequently appearing tokens and their frequency, i.e., the number of times that each occurs.   Please use the `simple_tokenize` function, without modification, to tokenize the text, so that everyone is working with the same definition of what a token is.   If you wish, feel free to start with the Python code in the box above - just copy it from there and paste it below. The return type should be a list of pairs (tuples) e.g. `[(word1, freq1), (word2, freq2), ...]` sorted in descending order by frequency (e.g. `top_50_tokens()[0]` should be the most common word and its frequency).

In [2]:
from simple_tokenize import simple_tokenize

def top_50_tokens(filename='Shakespeare.txt'):
    # In this function, write Python code to find the 50 most frequent tokens in Shakespeare.txt
    # Make sure that your code is commented
    dic = {}
    with open(filename) as f:
      for line in f:
        # tokenize, one line at a time
        words = simple_tokenize(line)
        for word in words:
          if word in dic:
            ## add the token frequency by one
            dic[word] += 1
          else:
            ## add the new token to the dictionary
            dic[word] = 1
    ## Sort the dictionary by frequencies
    sorted_dic = sorted(dic.items(), key = lambda x: x[1], reverse = True)
    ## Take the first 50 (key, value) pairs from the sorted dictionary
    top_50_tokens_list = sorted_dic[:50]
    return top_50_tokens_list


In [3]:
top_50_tokens()

[('the', 27378),
 ('and', 26082),
 ('i', 20717),
 ('to', 19661),
 ('of', 17473),
 ('a', 14723),
 ('you', 13630),
 ('my', 12490),
 ('in', 10996),
 ('that', 10915),
 ('is', 9137),
 ('not', 8512),
 ('with', 7778),
 ('me', 7777),
 ('it', 7692),
 ('for', 7578),
 ('be', 6867),
 ('his', 6859),
 ('your', 6657),
 ('this', 6606),
 ('but', 6277),
 ('he', 6260),
 ('have', 5885),
 ('as', 5744),
 ('thou', 5491),
 ('him', 5205),
 ('so', 5056),
 ('will', 4977),
 ('what', 4469),
 ('thy', 4034),
 ('all', 3923),
 ('her', 3850),
 ('no', 3797),
 ('by', 3768),
 ('do', 3753),
 ('shall', 3592),
 ('if', 3500),
 ('are', 3405),
 ('we', 3298),
 ('thee', 3180),
 ('on', 3062),
 ('lord', 3062),
 ('our', 3061),
 ('king', 2871),
 ('good', 2834),
 ('now', 2789),
 ('sir', 2763),
 ('from', 2640),
 ('o', 2621),
 ('come', 2519)]

Be sure to test the code that you have written by running it.   When you submit your notebook to us, we will run your code when we mark you assignment.   As a sanity test on you output, our reference implementation finds that the most frequent word is "the", which occurs 27378 times.

---

Once you have found the 50 most frequent tokens, let's move on to something slightly more complicated.

#### Question 3:

Instead of the most frequent tokens appearing in Shakespeare's works, suppose that we want a list of words that appear immediately after the word "perfect", on the same line, in Shakespeare's text. 
(Note: the "words" we are interested in for this question are tokens, as returned by simple_tokenize.)

For example, *All's Well That Ends Well* includes the line
>  Ere I can perfect mine intents, to kneel.

so "mine" should be part of the output, since it follows "perfect" on this line.  "intents", "to", and "kneel" should NOT be part of the output (at least from this line of text). To keep the output from getting too long, include only words that appear after "perfect" more than once. 

Notes: 
* If "perfect" is the last word on a line, then nothing follows it - remember, we only want words that are on the same line! 
* If "perfect" is followed by "perfect" then this counts! Don't forget to consider the word after the second perfect, too!

In the next box, implement the function `perfect_x` that returns a dictionary of key/value pairs, where the keys are the words that follow perfect on more than one line, and the values the number of lines in which the pattern is observed. For instance, if 'x' follows 'perfect' three times, the entry in the dictionary will be ('x':3) As a sanity check on your output, our reference implementation finds 5 such words.

In [4]:
from simple_tokenize import simple_tokenize

def perfect_x(filename = 'Shakespeare.txt'):
    perfect_x_dict = {}
    # In this function, write Python code to find tokens that follow "perfect" in Shakespeare.txt 
    # Make sure that your code is commented
    with open(filename) as f:
      for line in f:
        # tokenize, one line at a time
        words = simple_tokenize(line)
        ## add an indicator to see if perfect has been encountered.
        indicator = 0
        for word in words:
          if indicator == 1:
            ## add the token frequency by one
            perfect_x_dict[word] = perfect_x_dict.get(word, 0) + 1
            indicator = 0
          if word == 'perfect':
            indicator = 1
    key_to_del = []
    for k, v in perfect_x_dict.items():
      if v == 1:
        key_to_del.append(k)
    for k in key_to_del:
      del perfect_x_dict[k]
    return perfect_x_dict


In [5]:
perfect_x()

{'honour': 2, 'in': 4, 'love': 4, 'yellow': 2, 'that': 2}